backend/ringv2/ringv2.go

```diff
@@ -0,0 +1,433 @@
+		return "EventTrigger"
+	case EventError:
+		return "EventError"
+	case EventClosing:
+		return "EventClosing"
+	default:
+		return "INVALID"
+	}
+}
+
+const MinInterval = 5
+
+// Event represents an event that occurred in a ring. The event can originate
+// from any ring client.
+type Event struct {
+	// Type is the type of the event.
+	Type EventType
+
+	// Value is the ring item associated with the event.
+	Value string
+
+	// Err is any error that occurred while processing the event.
+	Err error
+}
+
+type Ring struct {
+	client *clientv3.Client
+
+	// itemPrefix is the prefix of all the ring items. The ring items are KVs
+	// that map from the value of the item to its sequence ID. The sequence
+	// IDs are updated on every ring advancement, and are used to sort the ring
+	// in order to find the least-valued item.
+	itemPrefix string
+
+	// keySeqKey is the key that points to the current sequence number of the
+	// ring. The sequence number is used to compute new keys within the ring
+	// prefix.
+	keySeqKey string
+
+	// intervalKey is the key that the TTL for ring items is stored at
+	intervalKey string
+
+	// triggerPrefix is the prefix that triggers are stored under. triggers are
+	// leased keys that are used to notify the ring clients about which item is
+	// next.
+	triggerPrefix string
+}
+
+// New creates a new Ring.
+func New(client *clientv3.Client, storePath string) *Ring {
+	return &Ring{
+		client:        client,
+		itemPrefix:    path.Join(storePath, "items"),
+		keySeqKey:     path.Join(storePath, "seq"),
+		intervalKey:   path.Join(storePath, "interval"),
+		triggerPrefix: path.Join(storePath, "triggers"),
+	}
+}
+
+// dump is for debugging
+func (r *Ring) dump(ctx context.Context, w io.Writer) {
+	resp, err := r.client.Get(ctx, path.Dir(r.itemPrefix), clientv3.WithPrefix())
+	if err != nil {
+		panic(err)
+	}
+	for _, kv := range resp.Kvs {
+		fmt.Fprintf(w, "%s: %s (%x)\n", string(kv.Key), string(kv.Value), kv.Value)
+	}
+}
+
+// IsEmpty returns true if there are no items in the ring.
+func (r *Ring) IsEmpty(ctx context.Context) (bool, error) {
+	resp, err := r.client.Get(ctx, r.itemPrefix,
+		clientv3.WithKeysOnly(),
+		clientv3.WithPrefix(),
+		clientv3.WithLimit(1))
+	if err != nil {
+		return false, err
+	}
+	return len(resp.Kvs) == 0, nil
+}
+
+func (r *Ring) grant(ctx context.Context) (*clientv3.LeaseGrantResponse, error) {
+	interval, err := r.getInterval(ctx)
+	if err != nil {
+		return nil, err
+	}
+	lease, err := r.client.Grant(ctx, interval)
+	return lease, err
+}
+
+// Add adds a new value to the ring. If the value already exists, it will not
+// be disturbed.
+func (r *Ring) Add(ctx context.Context, value string) error {
+	itemKey := path.Join(r.itemPrefix, value)
+
+	getresp, err := r.client.Get(ctx, itemKey)
+	if err != nil {
+		return fmt.Errorf("couldn't add %q to ring: %s", value, err)
+	}
+
+	if len(getresp.Kvs) > 0 && len(getresp.Kvs[0].Value) > 0 {
+		// Item already exists
+		return nil
+	}
```

Comment:
> If the value already exists, it will not be disturbed.

So this ring implementation is not a true container in the sense that it "inspects" values that are added to it and can take actions based on those values; in that case not add any "duplicates" to the ring.

This is fine for the original use case of the ring (round robin checks), but we also discussed using the new ring package for other things (to be determined!), implying it might need to be a bit more general and not assume that adding the same value several times is nonsensical.

What do you think?

Reply:
My view of the data structure is that it is a set - adding values should be idempotent. Additionally, the ordering should not be disturbed when an existing value is added.